src/nupic/encoders/RandomDistributedScalarEncoder.cpp

```diff
@@ -0,0 +1,149 @@
+  { initialize( parameters ); }
+
+void RandomDistributedScalarEncoder::initialize( const RDSE_Parameters &parameters)
+{
+  // Check parameters
+  NTA_CHECK( parameters.size > 0u );
+
+  UInt num_active_args = 0;
+  if( parameters.activeBits > 0u)   { num_active_args++; }
+  if( parameters.sparsity   > 0.0f) { num_active_args++; }
+  NTA_CHECK( num_active_args != 0u )
+      << "Missing argument, need one of: 'activeBits' or 'sparsity'.";
+  NTA_CHECK( num_active_args == 1u )
+      << "Too many arguments, choose only one of: 'activeBits' or 'sparsity'.";
+
+  UInt num_resolution_args = 0;
+  if( parameters.radius     > 0.0f) { num_resolution_args++; }
+  if( parameters.resolution > 0.0f) { num_resolution_args++; }
+  NTA_CHECK( num_resolution_args != 0u )
+      << "Missing argument, need one of: 'radius', 'resolution'.";
+  NTA_CHECK( num_resolution_args == 1u )
+      << "Too many arguments, choose only one of: 'radius', 'resolution'.";
+
+  args_ = parameters;
+  // Finish filling in all of parameters.
+
+  // Determine number of activeBits.
+  if( args_.sparsity > 0.0f ) {
+    NTA_CHECK( args_.sparsity >= 0.0f );
+    NTA_CHECK( args_.sparsity <= 1.0f );
+    args_.activeBits = (UInt) round( args_.size * args_.sparsity );
+    NTA_CHECK( args_.activeBits > 0u );
+  }
+  // Determine sparsity. Always calculate this even if it was given, to correct for rounding error.
+  args_.sparsity = (Real) args_.activeBits / args_.size;
+
+  // Determine resolution.
+  if( args_.radius > 0.0f ) {
+    args_.resolution = args_.radius / args_.activeBits;
+  }
+  // Determine radius.
+  else if( args_.resolution > 0.0f ) {
+    args_.radius = args_.activeBits * args_.resolution;
+  }
+
+  // Initialize parent class.
+  BaseEncoder<Real64>::initialize({ args_.size });
+}
+
+void RandomDistributedScalarEncoder::encode(Real64 input, sdr::SDR &output)
+{
+  // Check inputs
+  NTA_CHECK( output.size == size );
+  if( isnan(input) ) {
+    output.zero();
+    return;
+  }
+
+  auto &data = output.getDense();
+  fill( data.begin(), data.end(), 0u );
+
+  // Use the given seed to make a better, more randomized seed.
+  UInt32 apple_seed = MurmurHash3_x86_32(&args_.seed, sizeof(args_.seed), 0);
+
+  const UInt index = (UInt) (input / args_.resolution);
+  for(auto offset = 0u; offset < args_.activeBits; ++offset)
+  {
+    UInt hash_buffer = index + offset;
+    UInt32 bucket = MurmurHash3_x86_32(&hash_buffer, sizeof(hash_buffer), apple_seed);
+    bucket = bucket % size;
+
+    // Don't worry about hash collisions.  Instead measure the critical
+    // properties of the encoder in unit tests and quantify how significant
+    // the hash collisions are.  This encoder can not fix the collisions
+    // because it does not record past encodings.  Collisions cause small
+    // deviations in the sparsity or semantic similarity, depending on how
+    // they're handled.
+    // TODO: Calculate the probability of a hash collision and account for
```

Comment:
I was going to leave this as an exercise for the reader?

Reply:
> exercise for the reader?
 
never trust the user :) I think this could be a part of init checks in RDSE, tellining the user if #active bits is too low.

Reply:
There is a sanity check for `activeBits > 0`

Reply:
well, that does check only for basic usability, does not say anything about the collisions.. 
I''d suggest either implementing the check, if it's reasonably simple and generic. Or remove (just) the TODO , and keep the note "(You can) calculate the prob... "

Reply:
I removed the "TODO" note and changed it to explicitly say "Exercise for the reader:"